File: ml/app/schemas/divergence.py
"""Pydantic schemas for divergence detection endpoints."""

import datetime

from pydantic import BaseModel, Field


class DivergenceFeatureContribution(BaseModel):
    feature: str
    coefficient: float
    feature_value: float
    contribution: float
    direction: str  # "positive" | "negative"


class DivergenceDetectionResponse(BaseModel):
    date: str
    actual_score: float
    predicted_score: float
    residual: float
    cusum_positive: float = 0.0
    cusum_negative: float = 0.0
    cusum_alert: bool = False
    divergence_type: str = "aligned"
    confidence: float = Field(ge=0, le=1, default=0.0)
    top_drivers: list[DivergenceFeatureContribution] = []
    explanation: str = ""
    model_version: str = ""


class DivergenceRangeResponse(BaseModel):
    start: str
    end: str
    detections: list[DivergenceDetectionResponse]
    total_alerts: int
    model_version: str = ""


class DivergenceStatusResponse(BaseModel):
    is_ready: bool
    model_version: str = ""
    training_pairs: int = 0
    min_pairs_needed: int = 14
    r2_score: float | None = None
    mae: float | None = None
    phase: str = "cold_start"
    message: str = ""


class DivergenceTrainResponse(BaseModel):
    model_version: str
    training_pairs_used: int
    r2_score: float | None = None
    mae: float | None = None
    rmse: float | None = None
    message: str

File: api/domain/entity/divergence.go
package entity

import (
	"encoding/json"
	"time"
)

type DivergenceDetection struct {
	Date           time.Time       `json:"Date"`
	ConditionLogID int64           `json:"ConditionLogID"`
	ActualScore    float32         `json:"ActualScore"`
	PredictedScore float32         `json:"PredictedScore"`
	Residual       float32         `json:"Residual"`
	CuSumPositive  float32         `json:"CuSumPositive"`
	CuSumNegative  float32         `json:"CuSumNegative"`
	CuSumAlert     bool            `json:"CuSumAlert"`
	DivergenceType string          `json:"DivergenceType"`
	Confidence     float32         `json:"Confidence"`
	TopDrivers     json.RawMessage `json:"TopDrivers"`
	Explanation    string          `json:"Explanation"`
	ModelVersion   string          `json:"ModelVersion"`
	ComputedAt     time.Time       `json:"ComputedAt"`
}

type DivergenceModelStatus struct {
	IsReady        bool     `json:"IsReady"`
	ModelVersion   string   `json:"ModelVersion"`
	TrainingPairs  int      `json:"TrainingPairs"`
	MinPairsNeeded int      `json:"MinPairsNeeded"`
	R2Score        *float64 `json:"R2Score"`
	MAE            *float64 `json:"MAE"`
	Phase          string   `json:"Phase"`
	Message        string   `json:"Message"`
}

type DivergenceTrainResult struct {
	ModelVersion      string   `json:"ModelVersion"`
	TrainingPairsUsed int      `json:"TrainingPairsUsed"`
	R2Score           *float64 `json:"R2Score"`
	MAE               *float64 `json:"MAE"`
	RMSE              *float64 `json:"RMSE"`
	Message           string   `json:"Message"`
}
